src/common/paddle/organization.ts

```diff
@@ -0,0 +1,120 @@
+import { ContentPreferenceOrganization } from '../../entity/contentPreference/ContentPreferenceOrganization';
+import { ContentPreferenceStatus } from '../../entity/contentPreference/types';
+import { OrganizationMemberRoles } from '../../roles';
+import { SubscriptionProvider, SubscriptionStatus } from '../plus';
+import { logger } from '../../logger';
+import { isPlusMember } from '../../paddle';
+import { updateSubscriptionFlags } from '../utils';
+
+type PaddleCustomData = {
+  user_id?: string;
+};
+
+// If the user provides business information during the checkout, we will use it
+// otherwise we will use the default business name
+const getBusinessName = async (
+  user: User,
+  data: SubscriptionCreatedNotification,
+) => {
+  const defaultBusinessName = `${user.name}'s new Organization`;
+  if (!data.businessId) {
+    return defaultBusinessName;
+  }
+
+  try {
+    const business = await paddleInstance.businesses.get(
+      data.customerId,
+      data.businessId,
+    );
+    return business.name;
+  } catch (error) {
+    return defaultBusinessName;
+  }
+};
+
+export const createOrganizationSubscription = async ({
+  event,
+}: {
+  event: SubscriptionCreatedEvent;
+}) => {
+  const organizationId = randomUUID();
+  const con = await createOrGetConnection();
+
+  const { data } = event;
+  const { user_id: userId }: PaddleCustomData = data?.customData || {};
+
+  const subscriptionType = extractSubscriptionCycle(data.items);
+  if (!subscriptionType) {
+    logger.error(
+      {
+        provider: SubscriptionProvider.Paddle,
+        data: event,
+      },
+      'Subscription type missing in payload',
+    );
+    return false;
+  }
+
+  const user = await con.getRepository(User).findOneByOrFail({
+    id: userId,
+  });
+
+  const businessName = await getBusinessName(user, data);
+
+  const subscriptionFlags = {
+    cycle: subscriptionType,
+    createdAt: data.startedAt ?? undefined,
+    subscriptionId: data.id,
+    provider: SubscriptionProvider.Paddle,
+    status: SubscriptionStatus.Active,
+  };
+
+  const organization = await con.getRepository(Organization).save({
+    id: organizationId,
+    name: businessName,
+    seats: data.items[0].quantity,
+    subscriptionFlags: subscriptionFlags,
+  });
+
+  await Promise.all([
+    // Add the user to the organization
+    con.getRepository(ContentPreferenceOrganization).save({
+      userId: userId,
+      referenceId: organization.id,
+      organizationId: organization.id,
+      feedId: userId,
+      status: ContentPreferenceStatus.Follow,
+      flags: {
+        role: OrganizationMemberRoles.Owner,
+        referralToken: randomUUID(),
+      },
+    }),
+    // Update the paddle subscription with the organization id
+    // This is needed to be able to update the subscription later
+    paddleInstance.subscriptions.update(data.id, {
+      customData: {
+        user_id: userId,
+        organization_id: organization.id,
+      },
+    }),
```

Comment:
We choose not to do the frontend ID passing just so it can't fail?
But guess we can also match via user id if need be

Reply:
Yeah, just do it all on API side